File: backend/internal/auth/email_code.go
package auth

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"log"
	"net/http"
	"net/smtp"
	"os"
	"strings"
	"time"

	"practiq/internal/api"
	"practiq/internal/redisx"
)

const emailCodeTTL = 10 * time.Minute

type emailCodeRequest struct {
	Email *string `json:"email"`
}

// sendEmailCode is a package-level variable so tests can stub SMTP delivery.
var sendEmailCode = deliverEmailCode

// EmailCode issues a registration verification code to the given address.
func EmailCode() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body emailCodeRequest
		if err := decodeAuthRequest(w, r, &body); err != nil {
			api.HandleError(w, r, err)
			return
		}
		if detail := normalizeEmail(&body.Email, true); detail != nil {
			api.HandleError(w, r, api.ValidationError([]api.ValidationDetail{*detail}))
			return
		}
		if err := rateLimitAuth(r, *body.Email); err != nil {
			api.HandleError(w, r, err)
			return
		}
		rdb := redisx.Client()
		if rdb == nil {
			api.HandleError(w, r, sessionStoreUnavailable())
			return
		}
		code, err := newEmailCode()
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		if err := rdb.Set(r.Context(), emailCodeKey(*body.Email), code, emailCodeTTL).Err(); err != nil {
			api.HandleError(w, r, sessionStoreUnavailable())
			return
		}
		if err := sendEmailCode(*body.Email, code); err != nil {
			api.HandleError(w, r, api.NewError(http.StatusBadGateway, "EMAIL_DELIVERY_FAILED", "Could not send the verification email", nil))
			return
		}
		api.NoContent(w, r)
	})
}

// VerifyEmailCode checks the one-time registration code and consumes it on success.
func VerifyEmailCode(r *http.Request, email string, code string) error {
	invalid := api.ValidationError([]api.ValidationDetail{{Field: "code", Message: "Verification code is invalid or expired"}})
	if len(code) != 6 {
		return invalid
	}
	rdb := redisx.Client()
	if rdb == nil {
		return sessionStoreUnavailable()
	}
	key := emailCodeKey(email)
	stored, err := rdb.Get(r.Context(), key).Result()
	if err != nil || !hmac.Equal([]byte(stored), []byte(code)) {
		return invalid
	}
	_ = rdb.Del(r.Context(), key).Err()
	return nil
}

func emailCodeKey(email string) string {
	digest := sha256.Sum256([]byte(strings.ToLower(email)))
	return redisx.RedisKey("email-verify", hex.EncodeToString(digest[:]))
}

func newEmailCode() (string, error) {
	var b [4]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	n := uint32(b[0])<<24 | uint32(b[1])<<16 | uint32(b[2])<<8 | uint32(b[3])
	return fmt.Sprintf("%06d", n%1000000), nil
}

func deliverEmailCode(email string, code string) error {
	host := strings.TrimSpace(os.Getenv("SMTP_HOST"))
	if host == "" {
		// Local development fallback only; never leave SMTP_HOST unset in production.
		log.Printf("SMTP_HOST is not configured; verification code for %s is %s", email, code)
		return nil
	}
	port := strings.TrimSpace(os.Getenv("SMTP_PORT"))
	if port == "" {
		port = "587"
	}
	from := strings.TrimSpace(os.Getenv("SMTP_FROM"))
	username := strings.TrimSpace(os.Getenv("SMTP_USERNAME"))
	var auth smtp.Auth
	if username != "" {
		auth = smtp.PlainAuth("", username, os.Getenv("SMTP_PASSWORD"), host)
	}
	message := fmt.Sprintf("From: %s\r\nTo: %s\r\nSubject: PractiQ verification code\r\nMIME-Version: 1.0\r\nContent-Type: text/plain; charset=utf-8\r\n\r\nYour PractiQ verification code is %s. It expires in 10 minutes.\r\n", from, email, code)
	return smtp.SendMail(host+":"+port, auth, from, []string{email}, []byte(message))
}

File: server/auth/email_code.py
"""Email verification codes for registration (Redis + SMTP).

Mirrors backend/internal/auth/email_code.go.
"""

from __future__ import annotations

import hashlib
import hmac
import logging
import os
import secrets
import smtplib

from .. import envelope, redisx
from . import runtime

logger = logging.getLogger('practiq.auth')

EMAIL_CODE_TTL_SECONDS = 600


def _email_code_key(email: str) -> str:
    digest = hashlib.sha256(email.lower().encode()).hexdigest()
    return redisx.redis_key('email-verify', digest)


def _new_email_code() -> str:
    return f'{secrets.randbelow(1_000_000):06d}'


async def send_code(email: str) -> None:
    rdb = redisx.client()
    if rdb is None:
        raise runtime.session_store_unavailable()
    code = _new_email_code()
    try:
        await rdb.set(_email_code_key(email), code, ex=EMAIL_CODE_TTL_SECONDS)
    except Exception as exc:
        raise runtime.session_store_unavailable() from exc
    try:
        deliver_email_code(email, code)
    except Exception as exc:
        raise envelope.new_error(502, 'EMAIL_DELIVERY_FAILED', 'Could not send the verification email') from exc


async def verify_email_code(email: str, code: str) -> None:
    invalid = envelope.validation_error(
        [envelope.ValidationDetail('code', 'Verification code is invalid or expired')]
    )
    if len(code) != 6:
        raise invalid
    rdb = redisx.client()
    if rdb is None:
        raise runtime.session_store_unavailable()
    key = _email_code_key(email)
    stored = await redisx.get_text(rdb, key)
    if stored is None or not hmac.compare_digest(stored, code):
        raise invalid
    try:
        await rdb.delete(key)
    except Exception:
        pass


def deliver_email_code(email: str, code: str) -> None:
    host = os.environ.get('SMTP_HOST', '').strip()
    if not host:
        # Local development fallback only; never leave SMTP_HOST unset in production.
        logger.info('SMTP_HOST is not configured; verification code for %s is %s', email, code)
        return
    port = os.environ.get('SMTP_PORT', '').strip() or '587'
    from_address = os.environ.get('SMTP_FROM', '').strip()
    username = os.environ.get('SMTP_USERNAME', '').strip()
    password = os.environ.get('SMTP_PASSWORD', '')
    message = (
        f'From: {from_address}\r\nTo: {email}\r\nSubject: PractiQ verification code\r\n'
        'MIME-Version: 1.0\r\nContent-Type: text/plain; charset=utf-8\r\n\r\n'
        f'Your PractiQ verification code is {code}. It expires in 10 minutes.\r\n'
    )
    with smtplib.SMTP(host, int(port)) as smtp:
        if username:
            smtp.starttls()
            smtp.login(username, password)
        smtp.sendmail(from_address, [email], message.encode())
